plenum/server/backup_instance_faulty_processor.py

```diff
@@ -60,60 +60,62 @@ def process_backup_instance_faulty_msg(self,
         '''
         logger.debug("{} receive BackupInstanceFaulty "
                      "from {}: {}".format(self.node, frm, backup_faulty))
-
-        if getattr(backup_faulty, f.VIEW_NO.nm) != self.node.viewNo:
+        instances = getattr(backup_faulty, f.INSTANCES.nm)
+        if getattr(backup_faulty, f.VIEW_NO.nm) != self.node.viewNo or \
+                self.node.master_replica.instId in instances:
             return
 
         # Don't process BackupInstanceFaulty if strategy for this reason is not need quorum
         reason = Suspicions.get_by_code(getattr(backup_faulty, f.REASON.nm))
         if (
                 reason == Suspicions.BACKUP_PRIMARY_DISCONNECTED and
-                not self.__is_quorum_strategy(self.node.config.REPLICAS_REMOVING_WITH_PRIMARY_DISCONNECTED)
+                not self.is_quorum_strategy(self.node.config.REPLICAS_REMOVING_WITH_PRIMARY_DISCONNECTED)
         ) or (
                 reason == Suspicions.BACKUP_PRIMARY_DEGRADED and
-                not self.__is_quorum_strategy(self.node.config.REPLICAS_REMOVING_WITH_DEGRADATION)
+                not self.is_quorum_strategy(self.node.config.REPLICAS_REMOVING_WITH_DEGRADATION)
         ):
             return
 
         for inst_id in getattr(backup_faulty, f.INSTANCES.nm):
-            self.backup_instances_faulty.setdefault(inst_id, set()).add(frm)
             if inst_id not in self.node.replicas.keys():
                 continue
+            self.backup_instances_faulty.setdefault(inst_id, dict()).setdefault(frm, 0)
+            self.backup_instances_faulty[inst_id].setdefault(self.node.name, 0)
+            self.backup_instances_faulty[inst_id][frm] += 1
             if not self.node.quorums.backup_instance_faulty.is_reached(
-                    len(self.backup_instances_faulty[inst_id])):
+                    len(self.backup_instances_faulty[inst_id].keys())) \
+                    and not self.node.quorums.backup_instance_faulty.is_reached(
```

Comment:
1) Should we use a different quorum here?
2) Should we take into account only degradations from this node in a row, that is if we face degradation once per hour, it should not be accumulated, and replica should not be removed, right?

Reply:
As for the second item: should we send `BackupInstanceNotFaulty` when there is no degradation observed during the next check in monitor? `BackupInstanceNotFaulty` could clear all `backup_instance_faulty` for this node .
Unit tests need to be added for this.